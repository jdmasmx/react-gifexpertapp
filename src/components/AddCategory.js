import React, { useState } from 'react';
import PropTypes from 'prop-types';

//aqui podemos ver como setear el state de un componente padre
export const AddCategory = ({setCategories}) => {//con desestructuracion las props
    //export const AddCategory = (props) => {

    const [inputvalue, setinputvalue] = useState('');

    const handleInputChange = (e) => {
       setinputvalue( e.target.value );
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        //se llama el callback para referenciarlo
        if (inputvalue.trim().length > 2){
            //setCategories, es la funcion del componente padre
            //y se define el valor del state, y cats es el state que 
            //recibe es implicito para cada setXXX, en herencia si se 
            //define pero como en el componente padre, no es tan necesario definirle el state
            //
            
            setCategories( (cats) =>  [inputvalue,...cats]);
            setinputvalue( '' );//como aqui, no se define pero implicito que es inputvalue
        }       
    }

    return (
            <form onSubmit={ handleSubmit }>
                <input 
                    type="text" 
                    value={ inputvalue }
                    onChange={ handleInputChange }
                />
            </form>
        )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
