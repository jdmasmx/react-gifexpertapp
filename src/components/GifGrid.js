import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { getGifs } from '../helpers/getGifs';
 import { GifGridItem } from './GifGridItem';


export const GifGrid = ({ category }) => {

//ahora usamos el custom hook para hacer la peticion fetch
//y de aqui se manda la data, se usa la state del otro componente
//no de este como era en un inicio

const { data, loading } = useFetchGifs( category );

//const [images, setImages] = useState([])
//fue la 2da manera de consumirlo antes del custom hook, usando el helper aqui
//aun usabamos images,Setimages
//como tipo componentdidmount, y solo se ejecuta cuando es renderixado por primer avez el componente    
// useEffect(() => {
//     getGifs(category)//por que el helper es una promesa por el async
//         //.then( imgsrt => setImages ( imgsrt ))
//         .then( setImages )//es una funcion que el primer parametro viene implicito
// }, [ category ]);
//esta logica s eva al custom hook


//fue la 1ra manera antes del helper, se consumia directo aqui
// const getGifs = async() => {
//     const url = `https://api.giphy.com/v1/gifs/search?q=${ encodeURI( category ) }&limit=10&api_key=Vq8x04XtNrjIrcGNswBc3gQg7g3g6kQV`;
//     const resp = await fetch( url );
//     const { data } =  await resp.json();

//     const gifs = data.map(img => {
//         return {
//             id: img.id,
//             title: img.title,
//             url: img.images?.downsized_medium.url
//         }

//     })   
//     setImages( gifs );
   
// } 

    

  
    return (
        <>
        <h3 className='animate__animated animate__flash'>{category}</h3>
        { loading && <p className='flash'>Loading</p> }
         <div className='card-grid'>
                  {
                      data.map( img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img } />
                      ))
              }  
        </div>
        {/* <div className='card-grid'>
                  {//images, por que es el state
                      images.map( img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img } 
                            img2={img}/>
                      ))
              }  
        </div> */}
        </>
    )
}
