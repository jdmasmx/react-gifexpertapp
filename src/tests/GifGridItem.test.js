import React from 'react';
import { shallow } from "enzyme";
import { GifGridItem } from '../components/GifGridItem';



describe('Pruebas GifGridItem', () => { 

    const title = 'Un titulo';
    const url = 'https://algo.com/hola.jpg';
    const wrapper = shallow( <GifGridItem title={ title } url={ url} /> );

    test('Debe mostrar el componente correctamente', () => { 
        expect( wrapper ).toMatchSnapshot();
    })

    test('debe contener un parrafo con el title', () => { 

        const p = wrapper.find('p');
        expect( p.text().trim() ).toBe( title );

     })

     test('debe contener una imagen igual  a la url y alt de los props', () => { 

        const img = wrapper.find('img');
       //console.log( img.props() );//obtener todas las props
       //console.log( img.props().src );//obtener una en especifico de todas las props
       //obtener solo una prop
       //console.log( img.prop('src') );
       expect( img.prop('src') ).toBe( url );
       expect( img.prop('alt') ).toBe( title );

     })
     
     test('debe tener el animate__fadeIn', () => { 
        
        const div = wrapper.find('div');
        const className =  div.prop('className');
        //console.log(div.html());
        //console.log( div.props() );

        expect( className.includes('animate__fadeIn') ).toBe( true );


      })


 })