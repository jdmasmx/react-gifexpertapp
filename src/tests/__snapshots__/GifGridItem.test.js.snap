// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas GifGridItem Debe mostrar el componente correctamente 1`] = `
<div
  className="card animate__animated animate__bounce animate__fadeIn"
>
  <img
    alt="Un titulo"
    src="https://algo.com/hola.jpg"
  />
  <p>
    Un titulo
  </p>
</div>
`;
